lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -251,85 +217,64 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
+    -- many blocks, an error will be given.
     --
     -- Copied from the old logic but modified to use an accumulator rather
     -- than fmapping (<|). That changed the order so we're now NewestFirst
     -- (presumably the server sends them oldest first, as that assumption was
     -- required for the old version to correctly say OldestFirst).
-    --
-    -- FIXME it strikes me as weird that we check for hash equality to determine
-    -- when to stop. Why not stop when the server stops feeding?
     retrieveBlocksDo
-        :: Int        -- ^ Index of block we're requesting
-        -> ConversationActions MsgGetBlocks MsgBlock d
-        -> HeaderHash -- ^ We're expecting a child of this block
-        -> HeaderHash -- ^ Block at which to stop
+        :: ConversationActions MsgGetBlocks MsgBlock d
+        -> Int        -- ^ Index of block we're requesting
         -> [Block]    -- ^ Accumulator
-        -> ExceptT Text d (NewestFirst NE Block)
-    retrieveBlocksDo !i conv prevH endH !acc = lift (recvLimited conv) >>= \case
-        Nothing ->
-            throwError $ sformat ("Failed to receive block #"%int) i
-        Just (MsgNoBlock t) ->
-            throwError $ sformat ("Server failed to return block #"%int%": "%stext) i t
-        Just (MsgBlock block) -> do
-            let prevH' = block ^. prevBlockL
-                curH = headerHash block
-            when (prevH' /= prevH) $ do
-                throwError $ sformat
-                    ("Received block #"%int%" with prev hash "%shortHashF%
-                     " while "%shortHashF%" was expected: "%build)
-                    i prevH' prevH (block ^. blockHeader)
-            -- FIXME
-            -- Something to do with recording progress here.
-            -- Perhaps we could restore it by offering a streaming interface
-            -- for getBlocks.
-            --progressHeaderVar <- view (lensOf @ProgressHeaderTag)
-            --atomically $ do void $ tryTakeTMVar progressHeaderVar
-            --                putTMVar progressHeaderVar $ block ^. blockHeader
-            if curH == endH
-            -- Presumably the 'endH' block is the newest, i.e. the server
-            -- sends blocks oldest first.
-            then pure $ NewestFirst (block :| acc)
-            else retrieveBlocksDo (i+1) conv curH endH (block : acc)
-            --else over _Wrapped (block <|) <$> retrieveBlocksDo (i+1) conv curH endH
+        -> ExceptT Text d (NewestFirst [] Block)
+    retrieveBlocksDo conv !i !acc
+        | i <= 0    = pure $ NewestFirst acc
+        | otherwise = lift (recvLimited conv) >>= \case
+              Nothing ->
+                  throwError $ sformat ("Block retrieval cut short by peer at index #"%int) i
+              Just (MsgNoBlock t) ->
+                  throwError $ sformat ("Peer failed to produce block #"%int%": "%stext) i t
+              Just (MsgBlock block) -> do
+                  -- FIXME
```

Comment:
This must be resolved, other way daedalus will not ever sync. This variable is progress variable -- `m` in "synced `m/n`".

Reply:
This notion of progress is not a diffusion layer concern, so I'll put this back somewhere in the logic layer. In fact, it's so out of place at present. Merely *having* the block implies nothing about progress; we're concerned with how far along we are in block *application*. So I checked out the application pipeline and it's not what I expected. Apparently each list of `OldestFirst NE Blund` is applied piecewise: update payloads, then delegation payloads, then transaction payloads, then ssc payloads. Shame, it would surely be faster if we did one iteration over the batch of blocks, applying each piecewise update at each step.

But fixing that is out of scope here, so I just added yet another pass over the `OldestFirst NE Blund` to extract the newest block, and update the `ProgressHeaderVar` with its header hash (`applyBlocksDbUnsafeDo`). 6e5a556bfbd20934fd9779593072b0a318e4ec39